### PR TITLE
[backend] Process CSV keyword searching in the background

### LINKED ISSUE #7: [backend] Process CSV keyword searching in the background
## Why

The uploaded CSV keywords need to process in the background so that the app web process does not depend on it.

- Each keyword Google searching should be in the background
- Use Random agent while Google searching
- There should be some delay(prefer 1 to 5 seconds) between each search request so that Google does not block the IP
- Each keyword raw html response should be stored in the database and store the following information from the html response
   
   - Number of AdWords advertisers in the top position.
   - Total number of AdWords advertisers on the page.
   - URLs of the AdWords advertisers in the top position.
   - Number of the non-AdWords results on the page.
   - URLs of the non-AdWords results on the page.
   - Total number of links (all of them) on the page

 
## Who Benefits?

Users

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/elixir_google_scraper/projects/1#card-61312026

## What happened

Implemented Keyword scraping, parsing the response, storing  in the cron job with some delay
 
## Insight

- Used `httpoison` library for HTTP client
- Used `floki` library for html parsing
- Used `Oban` library for cron job
- Removed `async: true` from feature test as it facing race condition while running test

Note: There are some other refactoring needed for aligning with convention, I will do that at https://github.com/junan/elixir_google_scraper/issues/38

## Proof Of Work

Spec should be passed

### CODE REVIEW

File: lib/elixir_google_scraper/scraper.ex [hunk 1]
```diff
@@ -35,6 +42,20 @@ defmodule ElixirGoogleScraper.Scraper do
     end
   end
 
+  # There will be 3 seconds interval between each jonb starting time.
```
Comment (hanam1ni) [inline lib/elixir_google_scraper/scraper.ex]:
```suggestion
  # There will be 3 seconds interval between each job starting time.
```

File: lib/elixir_google_scraper/scraper/keyword_scraper.ex [hunk 2]
```diff
@@ -0,0 +1,30 @@
+defmodule ElixirGoogleScraper.Scraper.KeywordScraper do
+  alias ElixirGoogleScraper.Scraper.SearchResponseParser
+
+  def scrap(keyword) do
```
Comment (hanam1ni) [inline lib/elixir_google_scraper/scraper/keyword_scraper.ex]:
```suggestion
  def scrape(keyword) do
```

File: lib/elixir_google_scraper/scraper/keyword_scraper.ex [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  defp make_request(keyword) do
+    keyword
+    |> build_url
+    |> HTTPoison.get!(headers())
+  end
+
+  defp headers do
+    user_agent =
+      "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.447"
```
Comment (hanam1ni) [inline lib/elixir_google_scraper/scraper/keyword_scraper.ex]: Let's move this to module attribute

File: lib/elixir_google_scraper/scraper/worker/scraping_worker.ex [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    unique: [period: 30]
+
+  alias ElixirGoogleScraper.Repo
+  alias ElixirGoogleScraper.Scraper.{Keyword, KeywordScraper, SearchResult}
+
+  @impl Worker
+  def perform(%Oban.Job{args: %{"keyword_id" => keyword_id}}) do
+    keyword = Repo.get_by!(Keyword, %{id: keyword_id})
+
+    # Scaraping google search result and storing the parsed data
```
Comment (hanam1ni) [inline lib/elixir_google_scraper/scraper/worker/scraping_worker.ex]:
```suggestion
    # Scraping google search result and storing the parsed result
```

File: lib/elixir_google_scraper/scraper/worker/scraping_worker.ex [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    keyword.name
+    |> KeywordScraper.scrap()
+    |> SearchResult.changeset(%{keyword_id: keyword.id})
+    |> Repo.insert!()
+
+    # Updating keyword status as completed
+    keyword
+    |> Keyword.changeset(%{status: :completed})
+    |> Repo.update!()
```
Comment (hanam1ni) [inline lib/elixir_google_scraper/scraper/worker/scraping_worker.ex]: Let's move this logic to `keywords` context
```elixir
Keywords.mark_as_completed(keyword)
```

**Review by hanam1ni — COMMENTED** [review]
Sorry, 😢 I just notice one major refactor change.

According to our convention, the module `ElixirGoogleScraper.Scraper.Keyword` should split into two modules
- `ElixirGoogleScraper.Scraper.Keywords`: This is a **context module** for function related to keyword schema such as `mark_as_completed/1`
- `ElixirGoogleScraper.Scraper.Schema.Keyword`: This is a **schema module** for schema and changesets

For more detail, you can check here: https://nimblehq.co/compass/development/code-conventions/elixir-phoenix/#business-domain

You can fix in this PR or separate another PR for refactoring, let me know 🙌

**Comment by junan** [comment]
> Sorry, 😢 I just notice one major refactor change.

> According to our convention, the module `ElixirGoogleScraper.Scraper.Keyword` should split into two modules
> - `ElixirGoogleScraper.Scraper.Keywords`: This is a **context module** for function related to keyword schema such as `mark_as_completed/1`
> - `ElixirGoogleScraper.Scraper.Schema.Keyword`: This is a **schema module** for schema and changesets

> For more detail, you can check here: https://nimblehq.co/compass/development/code-conventions/elixir-phoenix/#business-domain

> You can fix in this PR or separate another PR for refactoring, let me know 🙌

Thanks @hanam1ni for pointing it out. I will refactor it in this PR 🙌

File: lib/elixir_google_scraper/scraper/keyword_scraper.ex [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  def scrape(keyword) do
+    response = make_request(keyword)
+    SearchResponseParser.parse(response.body)
+  end
+
+  defp google_search_url, do: "https://www.google.com/search"
+
+  defp make_request(keyword) do
+    keyword
+    |> build_url
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/keyword_scraper.ex]:
```suggestion
    |> build_url()
```

File: lib/elixir_google_scraper/scraper/keyword_scraper.ex [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  alias ElixirGoogleScraper.Scraper.SearchResponseParser
+
+  @user_agent "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.447"
+
+  def scrape(keyword) do
+    response = make_request(keyword)
+    SearchResponseParser.parse(response.body)
+  end
+
+  defp google_search_url, do: "https://www.google.com/search"
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/keyword_scraper.ex]: I think maybe we can move this after to `build_url` so we don't need to jump around too much. Or another approach is making this a constant on the top.

File: lib/elixir_google_scraper/scraper/search_result.ex [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      :top_ads_count,
+      :top_ads_urls,
+      :total_ads_count,
+      :result_count,
+      :result_urls,
+      :total_links_count,
+      :html,
+      :keyword_id
+    ])
+    |> validate_required([:result_count, :total_links_count, :html, :keyword_id])
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/search_result.ex]: We could add `|> assoc_constraint(:keyword)` to make sure the keyword with that id exists

File: test/elixir_google_scraper_web/features/upload_keywords/upload_keywords_test.exs [hunk 9]
```diff
@@ -1,5 +1,5 @@
 defmodule ElixirGoogleScraperWeb.UploadKeywords.UploadKeywordsTest do
-  use ElixirGoogleScraperWeb.FeatureCase, async: true
+  use ElixirGoogleScraperWeb.FeatureCase
```
Comment (rosle) [inline test/elixir_google_scraper_web/features/upload_keywords/upload_keywords_test.exs]: Just wondering why we're removing `async: true` from all feature cases? 🤔

File: test/elixir_google_scraper/scraper/worker/scraping_worker_test.exs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        assert search_result.total_links_count == 222
+        assert search_result.keyword_id == keyword.id
+        assert String.contains?(search_result.html, "<!doctype html>")
+      end
+    end
+
+    test "updates the keyword as completed" do
+      use_cassette "scraper/keyword_buy_bike" do
+        user = insert(:user)
+        keyword = insert(:keyword, name: "buy bike", user: user)
```
Comment (rosle) [inline test/elixir_google_scraper/scraper/worker/scraping_worker_test.exs]:
```suggestion
        keyword = insert(:keyword, name: "buy bike", user: user, status: :pending)
```

and maybe we can also remove user cause I think it doesn't matter here?

```suggestion
        keyword = insert(:keyword, name: "buy bike", status: :pending)
```

File: lib/elixir_google_scraper/scraper/keyword.ex [hunk 11]
```diff
@@ -19,4 +22,10 @@ defmodule ElixirGoogleScraper.Scraper.Keyword do
     |> cast(attrs, [:name, :status, :user_id])
     |> validate_required([:name, :user_id])
   end
+
+  def mark_as_completed(keyword) do
+    keyword
+    |> Keyword.changeset(%{status: :completed})
+    |> Repo.update!()
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/keyword.ex]: The `Repo.update!()` needs to be in context. Here we will just do changeset. we can have something like

```elixir
# Here
def complete_changeset(keyword) do
  change(keyword, %{status: :completed})
end

# In context
def mark_as_completed(keyword) do
  keyword
  |> complete_changeset()
  |> Repo.update!()
end
```

File: lib/elixir_google_scraper/scraper/schemas/keyword.ex [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  @doc false
+  def changeset(keyword, attrs) do
+    keyword
+    |> cast(attrs, [:name, :status, :user_id])
+    |> validate_required([:name, :user_id])
+  end
+
+  def complete_changeset(keyword) do
+    change(keyword, %{status: :completed})
+  end
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/schemas/keyword.ex:27]: We can also add unit tests for these schema changesets

File: lib/elixir_google_scraper/scraper/worker/scraping_worker.ex [hunk 13]
```diff
@@ -0,0 +1,27 @@
+defmodule ElixirGoogleScraper.Scraper.Worker.ScrapingWorker do
+  use Oban.Worker,
+    priority: 1,
+    max_attempts: 10,
+    unique: [period: 30]
+
+  alias ElixirGoogleScraper.Repo
+  alias ElixirGoogleScraper.Scraper.Keywords
+  alias ElixirGoogleScraper.Scraper.Schemas.Keyword
+  alias ElixirGoogleScraper.Scraper.{KeywordScraper, SearchResult}
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/worker/scraping_worker.ex]:
```suggestion
  alias ElixirGoogleScraper.Scraper.Schemas.Keyword
  alias ElixirGoogleScraper.Scraper.{Keywords, KeywordScraper, SearchResult}
```

File: lib/elixir_google_scraper/scraper/search_result.ex [hunk 14]
```diff
@@ -0,0 +1,37 @@
+defmodule ElixirGoogleScraper.Scraper.SearchResult do
```
Comment (rosle) [inline lib/elixir_google_scraper/scraper/search_result.ex]: Maybe moved to `/schemas/SearchResult.ex` like the `keyword` schema and add tests for changeset 👍

File: test/elixir_google_scraper_web/features/upload_keywords/upload_keywords_test.exs (hunk 9 above)
Comment (rosle) [reply test/elixir_google_scraper_web/features/upload_keywords/upload_keywords_test.exs]: I think we can leave it @junan, Later when we have time, we can test it again. Maybe you can create an issue.

Comment (junan) [reply test/elixir_google_scraper_web/features/upload_keywords/upload_keywords_test.exs]: @rosle I found the issue(https://github.com/parroty/exvcr/issues/58, https://github.com/parroty/exvcr/issues/53), it is related to `Exvcr` package that it sometime can't run tests parallel which it used a VCR cassette. I created a ticket for it to check it more later https://github.com/junan/elixir_google_scraper/issues/39. Currently one one test is nor running in parallel `scraping_worker_test.Exes` which using the VCR cassette

File: test/elixir_google_scraper/scraper/schemas/search_response_test.exs [hunk 15]
```diff
@@ -0,0 +1,39 @@
+defmodule ElixirGoogleScraper.Scraper.Schemas.SearchResultTest do
```
Comment (rosle) [inline test/elixir_google_scraper/scraper/schemas/search_response_test.exs]: the file name should be `test/elixir_google_scraper/scraper/schemas/search_result_test.exs` 👍

File: test/elixir_google_scraper/scraper/schemas/keyword_test.exs [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    test "requires the user_id" do
+      attrs = %{name: "shopping"}
+      changeset = Keyword.changeset(%Keyword{}, attrs)
+
+      refute changeset.valid?
+    end
+  end
+
+  describe "complete_changeset/1" do
+    test "updates the status from pending to completed in the changese" do
```
Comment (rosle) [inline test/elixir_google_scraper/scraper/schemas/keyword_test.exs]:
```suggestion
    test "updates the status from pending to completed in the changeset" do
```

**Review by hanam1ni — APPROVED** [review]

File: test/elixir_google_scraper/scraper/schemas/keyword_test.exs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+      refute changeset.valid?
+    end
+
+    test "requires the user_id" do
+      attrs = %{name: "shopping"}
+      changeset = Keyword.changeset(%Keyword{}, attrs)
+
+      refute changeset.valid?
+    end
```
Comment (rosle) [inline test/elixir_google_scraper/scraper/schemas/keyword_test.exs:23]: Normally for this, we can do in one test and maybe add a success case. Some thing like this

```elixir
test "returns valid changeset if given valid attributes" do
  attrs = ...[valid attrs]...

  changeset = Keyword.changeset(%Keyword{}, attrs)

  assert changeset.valid?
end

test "returns invalid changeset if required attributes are missing" do
  changeset = Keyword.changeset(%Keyword{}, %{})

  refute changeset.valid?
  
  assert errors_on(changeset) == %{
    ...
  }
end
```

For search result, we can have another one test case for assoc constraint.

```elixir
test "returns invalid changeset if keyword does not exist" do
  attrs = ...[attributes with non existing keyword id]...

  changeset = Keyword.changeset(%Keyword{}, attrs)

  assert {:error, changeset} = Repo.insert(changeset)
  refute changeset.valid?
  assert errors_on(changeset) == %{keyword: ["does not exist"]}
end
```

For assoc constraint the error is returning after db insert, so we need to call insert and check the result after 👍

**Review by rosle — APPROVED** [review]
👍

**Pull request merged**